app/data_client.go

```diff
@@ -761,3 +848,374 @@ func (d *DataClient) RemoveBinaryDataFromDatasetByIDs(
 	})
 	return err
 }
+
+func uploadMetadataToProto(metadata UploadMetadata) *syncPb.UploadMetadata {
+	methodParams, err := protoutils.ConvertMapToProtoAny(metadata.MethodParameters)
+	if err != nil {
+		return nil
+	}
+	return &syncPb.UploadMetadata{
+		PartId:           metadata.PartID,
+		ComponentType:    metadata.ComponentType,
+		ComponentName:    metadata.ComponentName,
+		MethodName:       metadata.MethodName,
+		Type:             syncPb.DataType(metadata.Type),
+		FileName:         metadata.FileName,
+		MethodParameters: methodParams,
+		FileExtension:    metadata.FileExtension,
+		Tags:             metadata.Tags,
+	}
+}
+
+func annotationsToProto(annotations Annotations) *pb.Annotations {
+	var protoBboxes []*pb.BoundingBox
+	for _, bbox := range annotations.Bboxes {
+		protoBboxes = append(protoBboxes, &pb.BoundingBox{
+			Id:             bbox.ID,
+			Label:          bbox.Label,
+			XMinNormalized: bbox.XMinNormalized,
+			YMinNormalized: bbox.YMinNormalized,
+			XMaxNormalized: bbox.XMaxNormalized,
+			YMaxNormalized: bbox.YMaxNormalized,
+		})
+	}
+	return &pb.Annotations{
+		Bboxes: protoBboxes,
+	}
+}
+
+func sensorMetadataToProto(metadata SensorMetadata) *syncPb.SensorMetadata {
+	return &syncPb.SensorMetadata{
+		TimeRequested: timestamppb.New(metadata.TimeRequested),
+		TimeReceived:  timestamppb.New(metadata.TimeReceived),
+		MimeType:      syncPb.MimeType(metadata.MimeType),
+		Annotations:   annotationsToProto(metadata.Annotations),
+	}
+}
+
+func sensorDataToProto(sensorData SensorData) *syncPb.SensorData {
+	protoSensorData := &syncPb.SensorData{
+		Metadata: sensorMetadataToProto(sensorData.Metadata),
+	}
+	switch {
+	case len(sensorData.SDBinary) > 0:
+		protoSensorData.Data = &syncPb.SensorData_Binary{
+			Binary: sensorData.SDBinary,
+		}
+	case sensorData.SDStruct != nil:
+		pbStruct, err := structpb.NewStruct(sensorData.SDStruct)
+		if err != nil {
+			return nil
+		}
+		protoSensorData.Data = &syncPb.SensorData_Struct{
+			Struct: pbStruct,
+		}
```

Comment:
(q) Is it possible that both `sensorData.SDBinary` will not be empty _and_ `sensorData.SDStruct` is not nil? If so, is preferencing `SDBinary` correct?

Reply:
so `dataCaptureUpload()` calls on `sensorDataToProto()` as a helper, and `dataCaptureUpload()` is called inside `binaryDataCaptureUpload()` or `tabularDataCaptureUpload()` and in both of those functions, only binary data or only tabular data is passed. It should never be the case that binary and tabular data are being uploaded together when this (`sensorDataToProto()`) helper function is called. But to your point, in the chance that is does get called with both it is prioritizing binaryData....tbh I don't know if that is the right call, maybe I should add a check to make sure that sensorData is only receiving binary or tabular data and otherwise, give an error? what do you think

Reply:
yeah, given the helper function where we only pass one, I think it's fine to just leave as is. Thanks for clarifying!